README

   gpart README


   Gpart is a small tool which tries to guess what partitions
   are on a PC type harddisk in case the primary partition table
   was damaged.

   Gpart works by scanning through the device (or file) given on
   the command line on a sector basis. Each guessing module is
   asked if it thinks a file system it knows about could start at
   a given sector. Several file system guessing modules are built
   in, others can be added dynamically.

   Consult the manual page for command line options and usage.



 - Installation ----------------------------------------------------------

   See file INSTALL.



 - Currently recognized partitions/filesystems types ---------------------

   Modname  Typ   Description
   fat      0x01  "Primary DOS with 12 bit FAT"
    	    0x04  "Primary DOS with 16 bit FAT (<= 32MB)"
            0x06  "Primary 'big' DOS (> 32MB)"
            0x0B  "DOS or Windows 95 with 32 bit FAT"
            0x0C  "DOS or Windows 95 with 32 bit FAT, LBA"
   ntfs     0x07  "OS/2 HPFS, NTFS, QNX or Advanced UNIX"
   hpfs     0x07  "OS/2 HPFS, NTFS, QNX or Advanced UNIX"
   ext2     0x83  "Linux ext2 filesystem"
   lswap    0x82  "Linux swap"
   bsddl    0xA5  "FreeBSD/NetBSD/386BSD"
   s86dl    0x82  "Solaris/x86 disklabel"
   minix    0x80  "Minix V1"
            0x81  "Minix V2"
   reiserfs 0x83  "ReiserFS filesystem"
   hmlvm    0xFE  "Linux LVM physical volumes"
   qnx4     0x4F  "QNX 4.x"
   beos     0xEB  "BeOS fs"
   xfs      0x83  "SGI XFS filesystem"



 - External guessing modules ---------------------------------------------

   gpart allows external partition type guessing modules to be
   added dynamically. An external module of type "xxx" must be
   compiled into a shared object file called "gm_xxx.so".

   External modules must provide three functions callable from
   gpart:

      int xxx_init(disk_desc *d,g_module *m)
         Initialisation function. Will be called before a scan.
         It should return the minimum number of bytes it wants
         to receive for a test. The module should set the
         description of the filesystem/partition type it handles
         in g_module.m_desc. If the filesystem/partition type
         included a partition table like first sector (like the
         *BSD disklabels do), the flag m_hasptbl should be set.
         Another flag is m_notinext which means the tested type
         cannot reside in a logical partition.

      int xxx_term(disk_desc *d)
         Termination/cleanup function, called after the scanning
         of the device has been done.

      int xxx_gfun(disk_desc *d,g_module *m)
         The actual guessing function, called from within the
         scan loop. It should test the plausibility of the
         given sectors, and return its guess in m->m_guess (a
         probability between 0 and 1). See existing modules
         for examples.

         The given file descriptor d->d_fd can be used for seeking
         and reading (see e.g. gm_ext2.c which tries to read
         the first spare superblock). If a module is convinced
         that it has found a filesystem/partition start it should
         fill in the assumed begin and size of the partition.

         The test performed should not be too pedantic, for
         instance it should not be relied upon that the file-
         system is clean/was properly unmounted. On the other
         hand too much tolerance leads to misguided guesses,
         so a golden middle way must be found.

   To create a shared object file from C source, compile the
   module via

      gcc -Wall -fPIC -shared gm_xxx.c -o gm_xxx.so

   Then install the shared object in a directory searched by
   the dynamic linker, or set LD_LIBRARY_PATH accordingly.
   The new external module can then be included by calling
   gpart like

      gpart -t xxx <other options>



 - Output explanation ----------------------------------------------------

   Here is a sample 'gpart -v' run on my first IDE hard disk
   (comments in brackets):

dev(/dev/hda) mss(512) chs(1232/255/63)(LBA) #s(19792080) size(9664mb)
[
   mss is the medium sector size, chs the geometry retrieved
   from the OS (or from the command line), #s is the total
   sector count.
]

Primary partition(1)
   type: 006(0x06)(Primary 'big' DOS (> 32MB)) (BOOT)
   size: 502mb #s(1028097) s(63-1028159)
   chs:  (0/1/1)-(63/254/63)d (0/1/1)-(63/254/63)r
   hex:  80 01 01 00 06 FE 3F 3F 3F 00 00 00 01 B0 0F 00
[
   size: the size of the partition in megabytes, number of
   sectors and the sector range.
   chs: the partition table chs range (d) and the real one.
   If the number of cylinders is less than 1024 both are
   identical.
   hex: the hexadecimal representation of the partition entry
   as found in the partition table.
]

...

Begin scan...
Possible partition(DOS FAT), size(502mb), offset(0mb)
   type: 006(0x06)(Primary 'big' DOS (> 32MB))
   size: 502mb #s(1028097) s(63-1028159)
   chs:  (0/1/1)-(63/254/63)d (0/1/1)-(63/254/63)r
   hex:  00 01 01 00 06 FE 3F 3F 3F 00 00 00 01 B0 0F 00

Possible extended partition at offset(502mb)
   Possible partition(Linux ext2), size(31mb), offset(502mb)
      type: 131(0x83)(Linux ext2 filesystem)
      size: 31mb #s(64196) s(1028223-1092418)
      chs:  (64/1/1)-(67/254/62)d (64/1/1)-(67/254/62)r
      hex:  00 01 01 40 83 FE 3E 43 7F B0 0F 00 C4 FA 00 00

   Possible partition(Linux swap), size(125mb), offset(533mb)
      type: 130(0x82)(Linux swap or Solaris/x86)
      size: 125mb #s(256976) s(1092483-1349458)
      chs:  (68/1/1)-(83/254/62)d (68/1/1)-(83/254/62)r
      hex:  00 01 01 44 82 FE 3E 53 83 AB 10 00 D0 EB 03 00
[
   During the scan phase all found partitions are listed by
   their real type names. The Linux swap partition above is
   recognized as Linux swap but will get the 0x82 partition
   identifier which can be both a Solaris disklabel or a
   Linux swap partition.

   When examing the hex values of the first primary partition
   it can be seen that they are identical to the values of the
   actual partition table (good guess) except for the first
   value (0x80 vs. 0x00). This entry denotes the partition
   'boot' flag which cannot be guessed.
]
...
End scan.

Checking partitions...
Partition(Primary 'big' DOS (> 32MB)): primary
   Partition(Linux ext2 filesystem): logical
   Partition(Linux swap or Solaris/x86): logical
   Partition(Linux LVM physical volume): logical
   Partition(Linux ext2 filesystem): logical
   Partition(DOS or Windows 95 with 32 bit FAT, LBA): logical
Partition(FreeBSD/NetBSD/386BSD): primary
Partition(Linux LVM physical volume): primary
Ok.
[
   During the scan phase gpart gathers a simple list of possible
   partitions, the check phase now tries to decide if found
   extended partitions seem consistent, if partitions do not
   overlap etc. Overlapping partitions are silently discarded,
   all remaining ones are given an attribute 'primary', 'logical',
   'orphaned' or 'invalid'. If gpart is called like 'gpart -vv ...',
   it also tells why it thinks a partition guess is invalid.

   If any inconsistencies are found, gpart prints the number
   of remaining inconsistencies, otherwise it says 'Ok.'
]

Guessed primary partition table:
Primary partition(1)
   type: 006(0x06)(Primary 'big' DOS (> 32MB))
   size: 502mb #s(1028097) s(63-1028159)
   chs:  (0/1/1)-(63/254/63)d (0/1/1)-(63/254/63)r
   hex:  00 01 01 00 06 FE 3F 3F 3F 00 00 00 01 B0 0F 00

Primary partition(2)
   type: 005(0x05)(Extended DOS)
   size: 6157mb #s(12611025) s(1028160-13639184)
   chs:  (64/0/1)-(848/254/63)d (64/0/1)-(848/254/63)r
   hex:  00 00 01 40 05 FE FF 50 40 B0 0F 00 D1 6D C0 00

Primary partition(3)
   type: 165(0xA5)(FreeBSD/NetBSD/386BSD)
   size: 1396mb #s(2859570) s(13639185-16498754)
   chs:  (849/0/1)-(1023/254/63)d (849/0/1)-(1026/254/63)r
   hex:  00 00 C1 51 A5 FE FF FF 11 1E D0 00 32 A2 2B 00

Primary partition(4)
   type: 254(0xFE)(Linux LVM physical volume)
   size: 1608mb #s(3293325) s(16498755-19792079)
   chs:  (1023/254/63)-(1023/254/63)d (1027/0/1)-(1231/254/63)r
   hex:  00 FE FF FF FE FE FF FF 43 C0 FB 00 8D 40 32 00
[
   This is a resulting primary partition table. Note that
   the logical partition guesses were only used to create
   the extended partition entry. Up to now gpart cannot
   reconstruct a damaged logical partition chain itself.

   If a guessed primary partition table should be written to
   some file or device the user must specify (via the '-W'
   option) which partition gets the active (bootable) one.
]



   gpart README, Aug 1999, Michail Brzitwa <mb@ichabod.han.de>


Changes
v0.2.1
  - Fix distribution issues that prevented builds of dist tarball

v0.2
  - Import some FreeBSD patches
  - Default scan by sector size to handle new 4k alignment method
  - Properly consider the real size of the disk to not invalidate partitions

v0.1i
  - Switch to autoconf/automake for cross-platform support
  - Integrate patches from distributions to fix build and overflow issues

v0.1h
 - New support for the following filesystems: BeOS, QNX 4.x & SGI XFS.
 - Updated Reiser filesystem support (Francis Devereux <francis@devereux.tc>).
 - Updated LVM support.
 - Several small fixes from contributors.

v0.1g
 - For access via raw devices: made writing of guessed table also aligned
   (reading has been aligned since 0.1f).
 - Fixed stupid copy&paste bug in the partition table check routine.

v0.1f
 - Default scan increment 'h' again.
 - Fixed wrong head-boundary condition.
 - Introduced possibility to edit guessed partitions.
 - Scan now starts on (sectors/head) unless -k <start> was given.
 - Length of guessed NTFS partitions now includes NTFS backup boot
   sector.

v0.1e
 - Default scan increment now 's', extended ptbl boundary condition
   now depends on scan increment (head if 's', else cylinder boundary).
 - Added LVM physical volume module (LVM by Heinz Mauelshagen).

v0.1d
 - Cope with short reads/read errors from disk (corrupted sectors etc.).
 - When flagging probable partitions 'invalid' give a reason why.
 - Fixed buggy ext2 spare superblock location calculation.

v0.1c
 - Don't flag partitions 'invalid' which are smaller than their
   partition table entry.
 - Fixed buggy rejection of some valid ext2 superblocks.
 - Fixed linux swap partition size calculation.
 - Added ReiserFS module.

v0.1b
 - First public version.
 - Made 'fast' scan the default behaviour (-f enables the complete scan).
 - Increased tolerance in recognizing extended partition tables.

v0.1a
 - Initial version.


src/gm_fat.c
/*      
 * gm_fat.c -- gpart fat guessing module
 *
 * gpart (c) 1999-2001 Michail Brzitwa <mb@ichabod.han.de>
 * Guess PC-type hard disk partitions.
 *
 * gpart is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2, or (at your
 * option) any later version.
 *
 * Created:   04.01.1999 <mb@ichabod.han.de>
 * Modified:
 *
 */

#include "gpart.h"
#include "gm_fat.h"

static const char	rcsid[] = "$Id: gm_fat.c,v 1.6 2001/02/07 18:08:08 mb Exp mb $";


int fat_init(disk_desc *d,g_module *m)
{
	if ((d == 0) || (m == 0))
		return (0);
	m->m_desc = "DOS FAT";
	m->m_align = 'h';
	return (sizeof(struct fat_boot_sector));
}



int fat_term(disk_desc *d)
{
        return (1);
}



int fat_gfun(disk_desc *d,g_module *m)
{
	struct fat_boot_sector	*sb = (struct fat_boot_sector *)d->d_sbuf;
	dos_part_entry		*pt = &m->m_part;
	unsigned long		nsecs = 0;
	unsigned char		ig1, ig2, media;
	int			sectsize, fat32 = 0, fat12 = 0;
	s64_t			size = 0;

	m->m_guess = GM_NO;
	ig1 = sb->ignored[0]; ig2 = sb->ignored[2]; media = sb->media;
	if ((ig1 == 0xeb) && (ig2 == 0x90) && ((media == 0xf8) || (media == 0xfc)))
	{
		if (*((unsigned short *)d->d_sbuf + 255) != le16(DOSPTMAGIC))
			return (1);

		/*
		 * looks like a standard FAT boot sector. Now find out,
		 * which one of the numerous versions this could be.
		 */

		pt->p_start = d->d_nsb;
		nsecs = le16(*(unsigned short *)sb->sectors);
		if (nsecs == 0)
			nsecs = le32(sb->total_sect);
		if (nsecs == 0)
			return (1);
		sectsize = le16(*(unsigned short *)sb->sector_size);
		if ((d->d_sbuf[0x39] == '1') && (d->d_sbuf[0x3a] == '2'))
			fat12 = 1;
		if (sb->fat_length == 0)
			fat32 = 1;
		if (fat32 && (*(d->d_sbuf + 0x26) == 0x29))
			return (1);
		if (!fat32 && (*(d->d_sbuf + 0x26) != 0x29))
			return (1);
		if (fat12 && fat32)
			return (1);
		m->m_guess = GM_YES;


		/*
		 * what happens when the fat sectsize != medium sectsize?
		 * I don't know. I just say no now.
		 */

		if (sectsize != d->d_ssize)
			m->m_guess = GM_NO;
		size = nsecs; size *= sectsize; size /= 1024;
		if (size >= 32768)
		{
			pt->p_typ = 0x06;
			if (fat32)
				pt->p_typ = d->d_lba ? 0x0C : 0x0B;
		}
		else
			pt->p_typ = fat12 ? 0x01 : 0x04;
		pt->p_size = nsecs;
	}
	return (1);
}
